src/core/compiler/CircomCompilerFactory.ts

```diff
@@ -1,44 +1,127 @@
-import fs from "fs";
+import os from "os";
+import path from "path";
+import fs from "fs-extra";
+import { promisify } from "util";
+import { exec } from "child_process";
 
-import { WASMCircomCompiler, NativeCircomCompiler } from "./CircomCompiler";
+import { Reporter } from "../../reporter";
 import { HardhatZKitError } from "../../errors";
-import { execCall } from "../../utils/utils";
+import { LATEST_SUPPORTED_CIRCOM_VERSION, OLDEST_SUPPORTED_ARM_CIRCOM_VERSION } from "../../constants";
 
-import { ICircomCompiler, IWASMCircomCompiler, CompilerVersion } from "../../types/core";
+import { BinaryCircomCompiler, WASMCircomCompiler } from "./CircomCompiler";
+import { CompilerDownloader } from "./CircomCompilerDownloader";
+
+import { isVersionHigherOrEqual } from "./versioning";
+
+import { CompilerInfo, CompilerPlatformBinary, ICircomCompiler, NativeCompiler } from "../../types/core";
 
 // eslint-disable-next-line
 const { Context } = require("@distributedlab/circom2");
 
 export class CircomCompilerFactory {
-  public static createWASMCircomCompiler(version: CompilerVersion): IWASMCircomCompiler {
-    switch (version) {
-      case "0.2.18":
-        return new WASMCircomCompiler(this._getCircomCompiler("@distributedlab/circom2/circom.wasm"));
-      default:
-        throw new HardhatZKitError(`Unsupported Circom compiler version - ${version}. Please provide another version.`);
+  public static async createBinaryCircomCompiler(version: string, isVersionStrict: boolean): Promise<ICircomCompiler> {
+    if (!isVersionHigherOrEqual(LATEST_SUPPORTED_CIRCOM_VERSION, version)) {
+      throw new HardhatZKitError(`Unsupported Circom compiler version - ${version}. Please provide another version.`);
     }
-  }
 
-  public static async createNativeCircomCompiler(version: CompilerVersion): Promise<ICircomCompiler> {
-    switch (version) {
-      case "0.2.18":
-        return new NativeCircomCompiler();
-      default:
-        throw new HardhatZKitError(`Unsupported Circom compiler version - ${version}. Please provide another version.`);
+    if (
+      isVersionStrict &&
+      os.arch() === "arm64" &&
+      !isVersionHigherOrEqual(version, OLDEST_SUPPORTED_ARM_CIRCOM_VERSION)
+    ) {
+      throw new HardhatZKitError(
+        `Circom compiler v${version} is not supported for ARM. Please provide a version not prior to ${OLDEST_SUPPORTED_ARM_CIRCOM_VERSION}.`,
+      );
+    }
+
+    const nativeCompiler = await this._getNativeCompiler();
+
+    if (nativeCompiler) {
+      const isValidVersion = isVersionStrict
+        ? nativeCompiler.version === version
+        : isVersionHigherOrEqual(nativeCompiler.version, version);
+
+      if (isValidVersion) {
+        Reporter!.reportCompilerVersion(nativeCompiler.version);
+
+        return new BinaryCircomCompiler(nativeCompiler.binaryPath);
+      }
+    }
+
```

Comment:
It is a bit strange that with the native compiler you use `if` and with other `try/catch`.